### PR TITLE
Enable nullability in some Control members

### PR DESCRIPTION
**Pull request by gpetrou**
## Proposed changes

- Enable nullability in some Control members.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/7144)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/SplitContainer.cs [hunk 1]
```diff
@@ -1555,13 +1555,13 @@ private int GetSplitterDistance(int x, int y)
         /// </summary>
         private bool ProcessArrowKey(bool forward)
         {
-            Control group = this;
+            Control? group = this;
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/SplitContainer.cs:1558]: How can this be nullable?

File: src/System.Windows.Forms/src/System/Windows/Forms/TabPage.cs [hunk 2]
```diff
@@ -621,7 +622,7 @@ protected override void OnPaintBackground(PaintEventArgs e)
             }
         }
 
-        internal override void RemoveToolTip(ToolTip toolTip)
+        internal override void RemoveToolTip(ToolTip? toolTip)
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/TabPage.cs:625]: There's only one call site (ignoring the tests), and that invocation passes `this`, so it can't be nullable. We can remove the check below and update the tests.

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs [hunk 3]
```diff
@@ -3027,7 +3028,7 @@ internal virtual Control ParentInternal
                     }
                     else
                     {
-                        _parent.Controls.Remove(this);
+                        _parent!.Controls.Remove(this);
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:3031]: Since `_parent` can be null, I don't think this is safe... 🤔 

I think we should update as follows:
```diff
diff --git a/src/System.Windows.Forms/src/System/Windows/Forms/Control.cs b/src/System.Windows.Forms/src/System/Windows/Forms/Control.cs
index c9119e0ca..076db89ea 100644
--- a/src/System.Windows.Forms/src/System/Windows/Forms/Control.cs
+++ b/src/System.Windows.Forms/src/System/Windows/Forms/Control.cs
@@ -3025,7 +3025,7 @@ namespace System.Windows.Forms
                     {
                         value.Controls.Add(this);
                     }
-                    else
+                    else if (_parent is not null)
                     {
                         _parent.Controls.Remove(this);
                     }
```

May worth reducing the nesting too, while we're at it:
```diff
diff --git a/src/System.Windows.Forms/src/System/Windows/Forms/Control.cs b/src/System.Windows.Forms/src/System/Windows/Forms/Control.cs
index c9119e0ca..5a37727dc 100644
--- a/src/System.Windows.Forms/src/System/Windows/Forms/Control.cs
+++ b/src/System.Windows.Forms/src/System/Windows/Forms/Control.cs
@@ -3019,19 +3019,21 @@ namespace System.Windows.Forms
             get => _parent;
             set
             {
-                if (_parent != value)
+                if (_parent == value)
                 {
+                    return;
+                }
+
                 if (value is not null)
                 {
                     value.Controls.Add(this);
                 }
-                    else
+                else if (_parent is not null)
                 {
                     _parent.Controls.Remove(this);
                 }
             }
         }
-        }
 ```

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/AxHostTests.cs [hunk 4]
```diff
@@ -2233,8 +2243,8 @@ public void AxHost_InvokeEditMode_InvokeWithSiteDesignModeWithComponentSelectedV
             Assert.Equal(0, invalidatedCallCount);
             Assert.Equal(0, styleChangedCallCount);
             Assert.Equal(0, createdCallCount);
-            mockSite.Verify(s => s.GetService(typeof(ISelectionService)), Times.Exactly(2));
-            mockSelectionService.Verify(s => s.GetComponentSelected(control), Times.Once());
+            mockSite.Verify(s => s.GetService(typeof(ISelectionService)), Times.Exactly(3));
+            mockSelectionService.Verify(s => s.GetComponentSelected(control), Times.Exactly(2));
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/AxHostTests.cs:2247]: Can you please help me understand what change led to this? Thank you.

**Review by RussKie — COMMENTED** [review]
👍 with few comments

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs (hunk 3 above)
Comment (gpetrou) [reply src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:3031]: We only enter the statement if `_parent` and `value` are different. We enter the `else` statement if `value` is null, which means that the `_parent` is not null. Do you still want me to add the extra check?

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/AxHostTests.cs (hunk 4 above)
Comment (gpetrou) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/AxHostTests.cs:2247]: If you try to debug these tests in current main, you will see that there are some NullReference exceptions that eventually get ignored. The problematic code seems to be `lpFrameInfo->hwndFrame = _host.ParentInternal.Handle;`. If we modify it to be like the code we have in WebBrowserSiteBase, we won't get these exceptions anymore, but we have to modify these values, so that the tests complete successfully.

**Comment by gpetrou** [comment]
@RussKie any more feedback on this?

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs (hunk 3 above)
Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:3031]: I'd add an extra check just for clarity.

**Review by RussKie — DISMISSED** [review]

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs [hunk 5]
```diff
@@ -885,7 +885,7 @@ protected virtual Point ScrollToControl(Control activeControl)
             if (activeControl.ParentInternal != this)
             {
                 Debug.WriteLineIf(s_autoScrolling!.TraceVerbose, "not direct child, original bounds: " + bounds);
-                bounds = RectangleToClient(activeControl.ParentInternal.RectangleToScreen(bounds));
+                bounds = RectangleToClient(activeControl.ParentInternal?.RectangleToScreen(bounds) ?? Rectangle.Empty);
```
Comment (dreddy-work) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs:888]: this change doesn't seem right. If you assign empty rectangle, rest of the code doesn't make sense. Can you explain why this change?

File: src/System.Windows.Forms/src/System/Windows/Forms/SplitContainer.cs [hunk 6]
```diff
@@ -1555,13 +1555,13 @@ private int GetSplitterDistance(int x, int y)
...
             if (ActiveControl is not null)
             {
                 group = ActiveControl.ParentInternal;
             }
 
-            return group.SelectNextControl(ActiveControl, forward, false, false, true);
+            return group?.SelectNextControl(ActiveControl, forward, false, false, true) ?? false;
```
Comment (dreddy-work) [inline src/System.Windows.Forms/src/System/Windows/Forms/SplitContainer.cs:1564]: it seems it was never null here before otherwise throws exception. Did you get chance to run the scenario here?

Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/SplitContainer.cs:1564]: `group` can be null, and we need to return a boolean.

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs (hunk 5 above)
Comment (gpetrou) [reply src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs:888]: `ParentInternal` is nullable. You want me to return `Point.Empty` instead if it is null?

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs (hunk 3 above)
Comment (gpetrou) [reply src/System.Windows.Forms/src/System/Windows/Forms/Control.cs]: I added the extra check.

File: src/System.Windows.Forms/src/System/Windows/Forms/UpDownBase.UpDownButtons.UpDownButtonsAccessibleObject.DirectionButtonAccessibleObject.cs [hunk 7]
```diff
@@ -44,7 +44,7 @@ public override Rectangle Bounds
                             }
 
                             // Convert to screen coords
-                            return (((UpDownButtons)_parent.Owner).ParentInternal).RectangleToScreen(bounds);
+                            return ((UpDownButtons?)_parent.Owner)?.ParentInternal?.RectangleToScreen(bounds) ?? Rectangle.Empty;
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/UpDownBase.UpDownButtons.UpDownButtonsAccessibleObject.DirectionButtonAccessibleObject.cs:47]: `_parent.Owner` can't be null here, we'd have failed on line:32 with a NRE.
```suggestion
                            return ((UpDownButtons)_parent.Owner).ParentInternal?.RectangleToScreen(bounds) ?? Rectangle.Empty;
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs (hunk 5 above)
Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs:888]: Should we check `activeControl.ParentInternal is null` at the beginning and bail out, if so?

Comment (gpetrou) [reply src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs]: I added a null check.

File: src/System.Windows.Forms/src/System/Windows/Forms/SplitContainer.cs (hunk 6 above)
Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Windows/Forms/SplitContainer.cs:1564]: I see that. Earlier it would have been throwing null reference if it was null.  But i don't think it ever did. So wanted to find scenario experience here If checking for null and throwing exception still make sense?

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs [hunk 8]
```diff
@@ -885,6 +885,12 @@ protected virtual Point ScrollToControl(Control activeControl)
             if (activeControl.ParentInternal != this)
             {
                 Debug.WriteLineIf(s_autoScrolling!.TraceVerbose, "not direct child, original bounds: " + bounds);
+
+                if (activeControl.ParentInternal is null)
+                {
+                    return Point.Empty;
```
Comment (dreddy-work) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs:891]: Same comment here. I am assuming, returning empty will place scroll bar back at (0,0) position. By understanding scenario here, ParentInternal may be null if tab goes out of scrollable control container. And in such case, scrollbars shouldn't reset to (0,0) position. @Tanya-Solyanik any thoughts here?

Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs:891]: Winforms invokes this method from a single place - https://github.com/dotnet/winforms/blob/0b2ee90e0c1c08458fdebaf72c81f0833e90b8f1/src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs#L841 where we are guaranteed that we don't enter this method when control in unparented because we test that control is a descendant. If user code called `ScrollToControl` internally, they were getting exception in the past, so we could throw a more descriptive exception now. So I would rather do `InvalidOperationException`.

File: src/System.Windows.Forms/src/System/Windows/Forms/TabPage.cs [hunk 9]
```diff
@@ -497,7 +497,8 @@ internal override void AssignParent(Control value)
         /// </summary>
         public static TabPage? GetTabPageOfComponent(object? comp)
         {
-            if (comp is not Control c)
+            Control? c = comp as Control;
+            if (c is null)
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/TabPage.cs:501]: Why change this?

File: src/System.Windows.Forms/src/Resources/SR.resx [hunk 10]
```diff
@@ -5194,6 +5194,9 @@ Stack trace where the illegal operation occurred was:
   <data name="ScrollableControlHorizontalScrollDescr" xml:space="preserve">
     <value>Gets the horizontal scroll bar for this ScrollableControl.</value>
   </data>
+  <data name="ScrollableControlParentInternalNull" xml:space="preserve">
+    <value>ParentInternal cannot be null.</value>
```
Comment (RussKie) [inline src/System.Windows.Forms/src/Resources/SR.resx:5198]: `ParentInternal` is an implementation detail, and it's not part of the public API surface.
```suggestion
    <value>Parent property of the active control cannot be null.</value>
```

similar to `DataGridViewCell_ValueTypeNull` and `DataGridViewCell_FormattedValueTypeNull`.

File: src/System.Windows.Forms/src/Resources/SR.resx [hunk 11]
```diff
@@ -5194,6 +5194,9 @@ Stack trace where the illegal operation occurred was:
...
+  </data>
```
Comment (RussKie) [inline src/System.Windows.Forms/src/Resources/SR.resx:5199]: `ParentInternal` is an implementation detail, and it's not part of the public API surface.

```suggestion
  <data name="ScrollableControlActiveControlParentNull" xml:space="preserve">
    <value>Parent property of the active control cannot be null.</value>
  </data>
```
similar to `DataGridViewCell_ValueTypeNull` and `DataGridViewCell_FormattedValueTypeNull`.

**Review by RussKie — COMMENTED** [review]
👍

File: src/System.Windows.Forms/src/System/Windows/Forms/TabPage.cs (hunk 9 above)
Comment (gpetrou) [reply src/System.Windows.Forms/src/System/Windows/Forms/TabPage.cs:501]: We need `c` to be nullable since we use `c.ParentInternal` below.

**Review by RussKie — APPROVED** [review]

**Comment by RussKie** [comment]
Thank you
